examples/meta-macro.kl

```diff
@@ -0,0 +1,30 @@
+#lang "n-ary-app.kl"
+
+(import (shift "n-ary-app.kl" 1))
+(import (shift "quasiquote.kl" 1))
+
+(define x 'foo)
+
+(define-macros
+  ((meta-macro1 (lambda (stx)
+                  (syntax-case stx
+                    ((list (_ macro-name var-name value))
+                     (pure `(define-macros
+                              ((,macro-name (lambda (stx)
+                                              (pure (list-syntax ('define ',var-name ',value) stx)))))))))))
+   (meta-macro2 (lambda (stx)
+                  (syntax-case stx
+                    ((list (_ macro-name))
+                     (pure `(define-macros
+                              ((,macro-name (lambda (stx)
+                                              (syntax-case stx
+                                                ((list (_ var-name value))
+                                                 (pure (list-syntax ('define var-name value) stx)))))))))))))))
+
+(meta-macro1 my-macro1 x 'bar)
+(my-macro1)
+(example x)  -- 'foo
```

Comment:
The fact that we get `'foo` here instead of `'bar` means that the `x` in the `meta-macro1` call is _not_ considered the same as the `x` in the `example` call. Is that the intended result? If find that counter-intuitive.

Reply:
This is the intended result. The idea is that references to names don't get to know about the internal implementation details of macros, including what they bind. In other words, the only thing that gets to refer to a macro-introduced binding is a macro-introduced use site. What you're intending here is to actually break the hygiene mechanism and do a bit of variable capture.

Here are two Racket programs that might be instructive. The first has the behavior that you're seeing in Klister:
```
#lang racket/base
(require (for-syntax racket/base racket/match))

(define x 'foo)

(define-syntax meta-macro1
  (lambda (stx)
    (match (syntax-e stx)
      [(list _ macro-name var-name value)
       #`(define-syntax #,macro-name
           (lambda (stx)
             (datum->syntax stx (list #'define #'#,var-name #'#,value))))])))

(meta-macro1 my-macro1 x 'bar)
(my-macro1)
x
```

The second has the behavior that you're wanting:
```
#lang racket/base
(require (for-syntax racket/base racket/match))

#;
(define x 'foo)

(define-syntax meta-macro1
  (lambda (stx)
    (match (syntax-e stx)
      [(list _ macro-name var-name value)
       #`(define-syntax #,macro-name
           (lambda (stx)
             (syntax-local-introduce (datum->syntax stx (list #'define #'#,var-name #'#,value)))))])))

(meta-macro1 my-macro1 x 'bar)
(my-macro1)
x
```

There are two differences:
 1. The second version comments out the `(define x 'foo)` because Racket doesn't allow top-level names to be duplicated.
2. The second invokes `syntax-local-introduce`, which breaks hygiene at one step of macro expansion.

`syntax-local-introduce` basically means "ignore hygiene right here". The way to implement it is to have it flip the current expansion scope. Then, the expander will flip it back when done. If we want it, and I think we do, then its type is `Syntax -> Macro Syntax`.